src/main/java/org/elasticsearch/index/query/TermQueryBuilder.java

```diff
@@ -147,7 +187,61 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
     }
 
     @Override
+    public Query toQuery(QueryParseContext parseContext) throws QueryParsingException, IOException {
+        Query query = null;
+        Preconditions.checkNotNull(this.fieldName, "Fieldname of a TermQuery cannot be null.");
+        Preconditions.checkNotNull(this.value, "Value of a TermQuery cannot be null.");
```

Comment:
First thing which I keep forgetting: you are right, queries can be created without going through `fromXContent`, java api will do just that. There are different places where we need the same validation, I think having a validate method is a good choice.

That said, one question is where do we call it from? `fromXContent` after parsing (on coordinating node in the future) makes perfect sense, but that is called only when the query is provided as json. Java api allows to create the intermediate query representation directly, meaning that the `fromXContent` step will be skipped there.  What can happen in this case is 1) that the query gets serialized to other nodes without being validated, and validated on the data nodes multiple times leading to multiple errors, which is not quite what we want (we want to catch errors once and earlier), 2) that the coordinating node is the data node where the query gets executed, query might not even get serialized, we execute `toQuery` straight-away.

I believe the proper way to validate the query is hooking into the existing request validation mechanism (`ActionRequest#validate`). Every request gets already validated on the coordinating node. The `SearchRequest` should simply call query#validate then I think (only problem being that we cannot quite do it yet, we need to refactor the search request first as right now the whole search request is just a big json object)? 
 
As for exceptions, if we adopt this approach I think we should be consistent with the existing validation infra and have `ActionRequestValidationException`s. 

Does this make sense to you guys?

If so, we need to decide what to do for now, I think having the validate method on the base class makes sense as Lee mentioned. For now we could call it from a single place though (`toQuery` is safer as it covers all the cases), once we move to parsing to the coord node we need to move it to the validation api instead.